<!DOCTYPE html>
<html lang="en"><head>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-b84e9dc6e68246e77ac4f52f8c844cc3.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.7">
<meta name="author" content="Dr.&nbsp;Cheng-Han Yu   Department of Mathematical and Statistical Sciences   Marquette University">
<title>MSSC 6250 - Spring 2025 – Tree-based Methods </title>
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
<link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
<link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
<style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #24292e;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #24292e; } /* Normal */
    code span.al { color: #ff5555; font-weight: bold; } /* Alert */
    code span.an { color: #6a737d; } /* Annotation */
    code span.at { color: #d73a49; } /* Attribute */
    code span.bn { color: #005cc5; } /* BaseN */
    code span.bu { color: #d73a49; } /* BuiltIn */
    code span.cf { color: #d73a49; } /* ControlFlow */
    code span.ch { color: #032f62; } /* Char */
    code span.cn { color: #005cc5; } /* Constant */
    code span.co { color: #6a737d; } /* Comment */
    code span.cv { color: #6a737d; } /* CommentVar */
    code span.do { color: #6a737d; } /* Documentation */
    code span.dt { color: #d73a49; } /* DataType */
    code span.dv { color: #005cc5; } /* DecVal */
    code span.er { color: #ff5555; text-decoration: underline; } /* Error */
    code span.ex { color: #d73a49; font-weight: bold; } /* Extension */
    code span.fl { color: #005cc5; } /* Float */
    code span.fu { color: #6f42c1; } /* Function */
    code span.im { color: #032f62; } /* Import */
    code span.in { color: #6a737d; } /* Information */
    code span.kw { color: #d73a49; } /* Keyword */
    code span.op { color: #24292e; } /* Operator */
    code span.ot { color: #6f42c1; } /* Other */
    code span.pp { color: #d73a49; } /* Preprocessor */
    code span.re { color: #6a737d; } /* RegionMarker */
    code span.sc { color: #005cc5; } /* SpecialChar */
    code span.ss { color: #032f62; } /* SpecialString */
    code span.st { color: #032f62; } /* String */
    code span.va { color: #e36209; } /* Variable */
    code span.vs { color: #032f62; } /* VerbatimString */
    code span.wa { color: #ff5555; } /* Warning */
  </style>
<link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto-89909b91675dd12fd7ae7f00f958249a.css">
<link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
<link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
<link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
<link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
<style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
<meta property="og:title" content="Tree-based Methods r fontawesome::fa(&quot;tree&quot;) – MSSC 6250 - Spring 2025">
<meta property="og:description" content="MSSC 6250 Statistical Machine Learning">
<meta property="og:site_name" content="MSSC 6250 - Spring 2025">
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-color="#447099" class="quarto-title-block center"><h1 class="title">Tree-based Methods <svg aria-hidden="true" role="img" viewbox="0 0 448 512" style="height:1em;width:0.88em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:currentColor;overflow:visible;position:relative;"><path d="M210.6 5.9L62 169.4c-3.9 4.2-6 9.8-6 15.5C56 197.7 66.3 208 79.1 208H104L30.6 281.4c-4.2 4.2-6.6 10-6.6 16C24 309.9 34.1 320 46.6 320H80L5.4 409.5C1.9 413.7 0 419 0 424.5c0 13 10.5 23.5 23.5 23.5H192v32c0 17.7 14.3 32 32 32s32-14.3 32-32V448H424.5c13 0 23.5-10.5 23.5-23.5c0-5.5-1.9-10.8-5.4-15L368 320h33.4c12.5 0 22.6-10.1 22.6-22.6c0-6-2.4-11.8-6.6-16L344 208h24.9c12.7 0 23.1-10.3 23.1-23.1c0-5.7-2.1-11.3-6-15.5L237.4 5.9C234 2.1 229.1 0 224 0s-10 2.1-13.4 5.9z"></path></svg>
</h1>
  <p class="subtitle">MSSC 6250 Statistical Machine Learning</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Dr.&nbsp;Cheng-Han Yu <br> Department of Mathematical and Statistical Sciences <br> Marquette University 
</div>
</div>
</div>

</section><section><section id="tree-based-methods" class="slide level2"><h2>Tree-based Methods</h2>
<ul>
<li><p>Can be used for regression and classification.</p></li>
<li><p>IDEA: <em>Segmenting the predictor space into many simple regions.</em></p></li>
<li><p>Simple, useful for interpretation, and has nice graphical representation.</p></li>
<li><p>Not competitive with the best supervised learning approaches in terms of prediction accuracy. (Large bias)</p></li>
<li><p>Combining a large number of trees (<strong>ensembles</strong>) often results in improvements in prediction accuracy, at the expense of some loss interpretation.</p></li>
</ul>
<!-- ## Adaptive  --><!-- # Classification and Regression Trees (CART) --></section><section id="decision-trees-classification-and-regression-trees-cart" class="slide level2"><h2>Decision Trees: Classification and Regression Trees (CART)</h2>
<ul>
<li><p>CART is a nonparametric method that <em>recursively partitions</em> the feature space into <em>hyper-rectangular</em> subsets (boxes), and make prediction on each subset.</p></li>
<li><p>Divide the predictor space — the set of possible values for <span class="math inline">\(X_1, X_2, \dots, X_p\)</span> — into <span class="math inline">\(J\)</span> distinct and non-overlapping regions, <span class="math inline">\(R_1, R_2, \dots, R_J\)</span>.</p></li>
</ul>

<img data-src="images/14-tree/unnamed-chunk-4-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:100.0%"><div class="fragment">
<ul>
<li>For every test point that falls into the region <span class="math inline">\(R_j\)</span> , we make the <span class="red">same</span> prediction:
<ul>
<li>
<em>Regression</em>: the <span class="green"><em>mean of the response values</em></span> for the training points in <span class="math inline">\(R_j\)</span>, i.e., <span class="math inline">\(\hat{y}_{R_j} = \sum_{k \in R_j} y_k / |R_j|\)</span>
</li>
<li>
<em>Classification</em>: the <span class="green"><em>most commonly occurring class</em></span> of training points in <span class="math inline">\(R_j\)</span>, i.e., <span class="math inline">\(\hat{y}_{R_j} = \underset{c}{\text{arg max}} \, \,\# (y_k = c)\)</span>, <span class="math inline">\(y_k \in R_j\)</span>.</li>
</ul>
</li>
</ul>
</div>
</section><section id="recursive-binary-splitting" class="slide level2"><h2>Recursive Binary Splitting</h2>
<ul>
<li>Computationally infeasible to consider every possible partition of the feature space into arbitrary <span class="math inline">\(J\)</span> boxes.</li>
</ul>
<div class="fragment">
<div class="columns">
<div class="column" style="width:80%;">
<ul>
<li>The <strong>recursive binary splitting</strong> is <em>top-down</em> and <em>greedy</em>:
<ul>
<li>
<span class="green">Top-down</span>: begins at the top of the tree (the entire <span class="math inline">\(X\)</span> space)</li>
<li>
<span class="green">Greedy</span>: at each step, the <em>best</em> split is made at that particular step, rather than looking ahead and picking a split that will lead to a better tree in some future step.</li>
</ul>
</li>
</ul>
</div><div class="column" style="width:20%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="./images/14-tree/8_1-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div></div>
</div>
<div class="fragment">
<ul>
<li>Select <span class="math inline">\(X_j\)</span> and a cutoff <span class="math inline">\(s\)</span> so that splitting the predictor space into <span class="math inline">\(\{\mathbf{X}\mid X_j &lt; s \}\)</span> and <span class="math inline">\(\{\mathbf{X}\mid X_j \ge s \}\)</span> leads to the greatest reduction in
<ul>
<li>
<span class="math inline">\(SS_{res}\)</span> for regression</li>
<li>
<strong>Gini index</strong>, <strong>entropy</strong> or misclassification rate for classification</li>
</ul>
</li>
<li>Repeatedly split one of the two previously identified regions until a stopping criterion is reached.</li>
</ul>
<aside class="notes"><p>at each step of the tree-building process</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section><section id="classification-tree" class="slide level2"><h2>Classification Tree</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li><p>KNN requires K and a distance measure.</p></li>
<li><p>SVM requires kernels.</p></li>
<li><p>Tree solves this by <em>recursively partitioning</em> the feature space using a <em>binary splitting</em> rule <span class="math inline">\(\mathbf{1}\{x \le c \}\)</span></p></li>
<li><p>0: <span class="red">Red</span>; 1: <span class="blue">Blue</span></p></li>
</ul>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-6-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section><section id="classification-tree-1" class="slide level2"><h2>Classification Tree</h2>
<p>If <span class="math inline">\(x_2 &lt; -0.64\)</span>, <span class="math inline">\(y = 0\)</span>.</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/rpart-plot-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-9-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section><section id="classification-tree-2" class="slide level2"><h2>Classification Tree</h2>
<p>If <span class="math inline">\(x_2 \ge -0.64\)</span> and <span class="math inline">\(x_1 \ge 0.69\)</span>, <span class="math inline">\(y = 0\)</span>.</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/rpart-plot-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-11-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section><section id="classification-tree-3" class="slide level2"><h2>Classification Tree</h2>
<p>If <span class="math inline">\(x_2 \ge -0.64\)</span>, <span class="math inline">\(x_1 &lt; 0.69\)</span>, and <span class="math inline">\(x_2 \ge 0.75\)</span>, <span class="math inline">\(y = 0\)</span>.</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/rpart-plot-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-13-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section><section id="classification-tree-4" class="slide level2"><h2>Classification Tree</h2>
<p>If <span class="math inline">\(x_2 \ge -0.64\)</span>, <span class="math inline">\(x_1 &lt; 0.69\)</span>, <span class="math inline">\(x_2 &lt; 0.75\)</span>, and <span class="math inline">\(x_1 &lt; -0.69\)</span>, <span class="math inline">\(y = 0\)</span>.</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/rpart-plot-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-15-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section><section id="classification-tree-5" class="slide level2"><h2>Classification Tree</h2>
<p>Step 5 may not be beneficial.</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/rpart-plot-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-17-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section><section id="classification-tree-6" class="slide level2"><h2>Classification Tree</h2>
<p>Step 6 may not be beneficial. (Could overfit)</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/rpart-plot-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-19-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section><section id="classification-tree-7" class="slide level2"><h2>Classification Tree</h2>
<p>Step 7 may not be beneficial. (Could overfit)</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/rpart-plot-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-21-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section><section id="misclassification-rate" class="slide level2"><h2>Misclassification Rate</h2>
<ul>
<li><p>The classification error rate is the fraction of the training observations in the region that do not belong to the most common class: <span class="math display">\[1 - \max_{k} (\hat{p}_{mk})\]</span> where <span class="math inline">\(\hat{p}_{mk}\)</span> is the proportion of training observations in the <span class="math inline">\(m\)</span>th region that are from the <span class="math inline">\(k\)</span>th class.</p></li>
<li><p>It is not sensitive for tree-growing.</p></li>
<li><p>Hope to have nodes (regions) including training points that belong to only one class.</p></li>
</ul></section><section id="gini-index-impurity" class="slide level2"><h2>Gini Index (Impurity)</h2>
<p>The Gini index is defined by</p>
<p><span class="math display">\[\sum_{k=1}^K \hat{p}_{mk}(1 - \hat{p}_{mk})\]</span> which is a measure of total variance across the <span class="math inline">\(K\)</span> classes.</p>
<ul>
<li><p>Gini is small if all of the <span class="math inline">\(\hat{p}_{mk}\)</span>s are close to zero or one.</p></li>
<li><p><strong>Node purity</strong>: a small value indicates that a node contains predominantly observations from a single class.</p></li>
</ul></section><section id="shannon-entropy" class="slide level2"><h2>Shannon Entropy</h2>
<p>The <strong>Shannon entropy</strong> is defined as</p>
<p><span class="math display">\[- \sum_{k=1}^K \hat{p}_{mk} \log(\hat{p}_{mk}).\]</span></p>
<ul>
<li><p>The entropy is near zero if the <span class="math inline">\(\hat{p}_{mk}\)</span>s are all near zero or one.</p></li>
<li><p>Gini index and the entropy are similar numerically.</p></li>
</ul>
<aside class="notes"><p>Any of these three approaches might be used when pruning the tree, but the classification error rate is preferable if prediction accuracy of the final pruned tree is the goal.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside></section><section id="comparing-measures" class="slide level2"><h2>Comparing Measures</h2>
<ul>
<li>Use Gini and Entropy for training (building a tree), and use error rate for evaluating predictive accuracy.</li>
</ul>
<!-- - Misclassification can be used for evaluating a tree, but may not be sensitive enough for building a tree. -->
<img data-src="images/14-tree/unnamed-chunk-22-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:100.0%"><aside class="notes"><p>For each quantity, smaller value means that the node is more “pure”, hence, there is a higher certainty when we predict a new value. The idea of splitting a node is that, we want the two resulting child node to contain less variation. In other words, we want each child node to be as “pure” as possible.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside></section><section id="regression-tree" class="slide level2"><h2>Regression Tree</h2>
<p>The goal is to find boxes <span class="math inline">\(R_1, \dots ,R_J\)</span> that minimize the <span class="math inline">\(SS_{res}\)</span>, given by <span class="math display">\[\sum_{j=1}^J\sum_{i \in R_j}\left( y_i - \hat{y}_{R_j}\right)^2\]</span> where <span class="math inline">\(\hat{y}_{R_j}\)</span> is the mean response for the training observations within <span class="math inline">\(R_j\)</span>.</p>
<div class="columns">
<div class="column" style="width:30%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="./images/14-tree/partition.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:70%;">
<div class="xsmall">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="./images/14-tree/reg-tree.png" style="width:80.0%"></p>
<figcaption>Source: ISL Fig 8.3</figcaption></figure>
</div>
</div>
</div>
</div>
</div></div>
</section><section id="tree-pruning" class="slide level2"><h2>Tree Pruning</h2>
<ul>
<li>Using regression and classification performance measures to grow trees with no penalty on the tree size leads to overfitting.</li>
</ul>
<div class="fragment">
<ul>
<li>
<strong>Cost complexity pruning</strong>:</li>
</ul>
<p>Given the largest tree <span class="math inline">\(T_{max}\)</span>,</p>
<p><span class="math display">\[\begin{align}
\min_{T \subset T_{max}} \sum_{m=1}^{|T|}\sum_{i:x_i\in R_m} \left( y_i - \hat{y}_{R_m}\right)^2 +  \alpha|T|
\end{align}\]</span> where <span class="math inline">\(|T|\)</span> indicates the number of terminal nodes of the tree <span class="math inline">\(T\)</span>.</p>
<ul>
<li><p>Large <span class="math inline">\(\alpha\)</span> results in small trees</p></li>
<li><p>Choose <span class="math inline">\(\alpha\)</span> using CV</p></li>
<li><p>Algorithm 8.1 in ISL for building a regression tree.</p></li>
<li><p>For classification, replace <span class="math inline">\(SS_{res}\)</span> with a classification performance measure.</p></li>
</ul>
</div>
</section><section id="implementation" class="slide level2"><h2>Implementation</h2>
<ul>
<li><a href="https://cran.r-project.org/web/packages/rpart/vignettes/longintro.pdf"><code>rpart::rpart()</code></a></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bethatkinson/rpart">rpart</a></span><span class="op">)</span></span>
<span><span class="fu">rpart</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rpart/man/rpart.html">rpart</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="va">x1</span> <span class="op">+</span> <span class="va">x2</span>, <span class="va">data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<ul>
<li><a href="https://cran.r-project.org/web/packages/tree/tree.pdf"><code>tree::tree()</code></a></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="fu">tree</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tree/man/tree.html">tree</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="va">x1</span> <span class="op">+</span> <span class="va">x2</span>, <span class="va">data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<ul>
<li><a href="https://scikit-learn.org/stable/modules/tree.html#"><code>sklearn tree</code></a></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn <span class="im">import</span> tree</span>
<span id="cb3-2"><a aria-hidden="true" tabindex="-1"></a>dtc <span class="op">=</span> tree.DecisionTreeClassifier()</span>
<span id="cb3-3"><a aria-hidden="true" tabindex="-1"></a>dtc <span class="op">=</span> dtc.fit(X, y)</span>
<span id="cb3-4"><a aria-hidden="true" tabindex="-1"></a>dtr <span class="op">=</span> tree.DecisionTreeRegressor()</span>
<span id="cb3-5"><a aria-hidden="true" tabindex="-1"></a>dtr <span class="op">=</span> dtr.fit(X, y)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="trees-v.s.-linear-regression" class="slide level2"><h2>Trees v.s. Linear Regression</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><span class="green"><strong>Linear regression</strong></span></p>
<p><span class="math display">\[f(X) = \beta_0 + \sum_{j=1}^pX_j\beta_j\]</span></p>
<ul>
<li>Performs better when the relationship between <span class="math inline">\(y\)</span> and <span class="math inline">\(x\)</span> is approximately linear.</li>
</ul>
</div><div class="column" style="width:50%;">
<p><span class="green"><strong>Regression tree</strong></span></p>
<p><span class="math display">\[f(X) = \sum_{j=1}^J \hat{y}_{R_j}\mathbf{1}(\mathbf{X}\in R_j)\]</span></p>
<ul>
<li><p>Performs better when there is a highly nonlinear and complex relationship between <span class="math inline">\(y\)</span> and <span class="math inline">\(x\)</span>.</p></li>
<li><p>Preferred for interpretability and visualization.</p></li>
</ul>
</div></div>
</section><section id="trees-v.s.-linear-models" class="slide level2"><h2>Trees v.s. Linear Models</h2>
<div class="xsmall">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="./images/14-tree/8_7-1.png" style="width:50.0%"></p>
<figcaption>Source: ISL Fig 8.7</figcaption></figure>
</div>
</div>
</div>
</div>
<!-- # Bagging {background-color="#447099"} -->
</section></section><section><section id="ensemble-learning-bagging-random-forests-boosting" class="title-slide slide level1 center"><h1>Ensemble Learning: Bagging, Random Forests, Boosting</h1>
<blockquote>
<p><strong>Two heads are better than one</strong>, not because either is infallible, but because they are unlikely to go wrong in the same direction. – C.S. Lewis, British Writer (1898 - 1963)</p>
</blockquote>
<blockquote>
<p>『三個臭皮匠，勝過一個諸葛亮』</p>
</blockquote>
</section><section id="ensemble-methods" class="slide level2"><h2>Ensemble Methods</h2>
<ul>
<li><p>An <strong>ensemble</strong> method combines many <strong>weak learners</strong> (unstable, less accurate) to obtain a single and powerful model.</p></li>
<li><p>The CARTs suffer from <em>high variance</em>.</p></li>
<li><p>If independent <span class="math inline">\(Z_1, \dots, Z_n\)</span> have variance <span class="math inline">\(\sigma^2\)</span>, then <span class="math inline">\(\bar{Z}\)</span> has variance <span class="math inline">\(\sigma^2/n\)</span>.</p></li>
<li><p><em>Averaging a set of observations reduces variance!</em></p></li>
</ul>
<div class="fragment">
<p>With <span class="math inline">\(B\)</span> separate training sets,</p>
<p><span class="math display">\[\hat{f}_{avg}(x) = \frac{1}{B}\sum_{b=1}^B\hat{f}_{b}(x)\]</span></p>
</div>
</section><section id="bagging" class="slide level2"><h2>Bagging</h2>
<div class="columns">
<div class="column" style="width:40%;">
<ul>
<li>
<strong>Bootstrap aggregation</strong>, or <strong>bagging</strong> is a procedure for reducing variance.</li>
</ul>
<!-- - Generate $B$ bootstrap samples from the training set. --><ul>
<li>Generate <span class="math inline">\(B\)</span> bootstrap samples by repeatedly sampling with replacement from the training set <span class="math inline">\(B\)</span> times.</li>
</ul>
<p><span class="math display">\[\hat{f}_{bag}(x) = \frac{1}{B}\sum_{b=1}^B\hat{f}^*_{b}(x)\]</span></p>
</div><div class="column" style="width:60%;">
<div class="xsmall">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="https://upload.wikimedia.org/wikipedia/commons/c/c8/Ensemble_Bagging.svg" style="width:100.0%"></p>
<figcaption>Source: Wiki page of bootstrap aggregating</figcaption></figure>
</div>
</div>
</div>
</div>
</div></div>
<aside class="notes"><p>Bagging works for many regression methods, but it is particularly useful for decision trees.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside></section><section id="bagging-on-decision-trees" class="slide level2"><h2>Bagging on Decision Trees</h2>
<div class="xsmall">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="https://upload.wikimedia.org/wikipedia/commons/4/4a/Ensemble_Aggregation.png" style="width:100.0%"></p>
<figcaption>Source: Wiki page of ensemble learning</figcaption></figure>
</div>
</div>
</div>
</div>
</section><section id="cart-v.s.-bagging" class="slide level2"><h2>CART v.s. Bagging</h2>
<ul>
<li><p>For CART, the decision line has to be aligned to axis.</p></li>
<li><p>For Bagging, <span class="math inline">\(B = 200\)</span> each having 400 training points. Boundaries are smoother.</p></li>
</ul>
<div class="columns">
<div class="column" style="width:33%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-37-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:33%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/cart-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:33%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/bagging-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div></div>
<!-- # Random Forests {background-color="#447099"} -->
</section><section id="notes-of-bagging" class="slide level2"><h2>Notes of Bagging</h2>
<ul>
<li><p>Using a large <span class="math inline">\(B\)</span> will not lead to overfitting.</p></li>
<li><p>Use <span class="math inline">\(B\)</span> sufficiently large that the error has settled down.</p></li>
<li><p>Bagging improves prediction accuracy at the expense of interpretability.</p></li>
</ul>
<div class="fragment">
<ul>
<li>
<p>When different trees are highly correlated, simply averaging is not very effective.</p>
<ul>
<li>If there is one very strong predictor, in the collection of bagged trees, all of the trees will use this strong predictor in the top split. Therefore, all of the bagged trees will look similar to each other.</li>
</ul>
</li>
<li><p>The predictions from the bagged trees will be highly correlated, and hence averaging does not lead to as large reduction in variance.</p></li>
</ul>
<aside class="notes"><ul>
<li><p>Use many trees and averaging. It is no longer clear which variables are most important to the procedure.</p></li>
<li><p>If there is one very strong predictor in the data set, in the collection of bagged trees, most or all of the trees will use this strong predictor in the top split. Therefore, all of the bagged trees will look quite similar to each other.</p></li>
<li><p>The predictions from the bagged trees will be highly correlated, and hence averaging does not lead to as large reduction in variance.</p></li>
</ul><style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section><section id="random-forests" class="slide level2"><h2>Random Forests</h2>
<div class="columns">
<div class="column" style="width:40%;">
<ul>
<li><p><strong>Random forests</strong> improve bagged trees by <em>decorrelating</em> the trees.</p></li>
<li><p><em><span class="math inline">\(m\)</span> predictors are randomly sampled as split candidates from the <span class="math inline">\(p\)</span> predictors.</em></p></li>
</ul>
<!-- - The split is allowed to use only one of those $m$ predictors. -->
</div><div class="column" style="width:60%;">
<div class="xsmall">
<!-- # ```{r} -->
<!-- # #| fig-cap: "Source: Misra and Li in Machine Learning for Subsurface Characterization (2020)" -->
<!-- # #| out-width: 50% -->
<!-- # knitr::include_graphics("./images/14-tree/rf.jpg") -->
<!-- # ``` -->
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="./images/14-tree/rf1.png" style="width:90.0%"></p>
<figcaption>Source: Multivariate Statistical Machine Learning Methods for Genomic Prediction, Lopez et al.&nbsp;(2022)</figcaption></figure>
</div>
</div>
</div>
</div>
</div></div>
<aside class="notes"><ul>
<li><p>When building decision trees, each time a split in a tree is considered, <em>a random sample of <span class="math inline">\(m\)</span> predictors is chosen as split candidates from the full set of <span class="math inline">\(p\)</span> predictors.</em></p></li>
<li><p>The split is allowed to use only one of those <span class="math inline">\(m\)</span> predictors.</p></li>
</ul><style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside></section><section id="random-forests-1" class="slide level2"><h2>Random Forests</h2>
<ul>
<li><p><span class="math inline">\(m \approx \sqrt{p}\)</span> for classification; <span class="math inline">\(m \approx p/3\)</span> for regression.</p></li>
<li><p><em>Decorrelating</em>: on average <span class="math inline">\((p − m)/p\)</span> of the splits will not even consider the strong predictor, and so other predictors will have more of a chance.</p></li>
<li><p>If <span class="math inline">\(m = p\)</span>, random forests = bagging.</p></li>
<li><p>The improvement is significant when <span class="math inline">\(p\)</span> is large.</p></li>
</ul></section><section id="cart-vs.-bagging-vs.-random-forests" class="slide level2"><h2>CART vs.&nbsp;Bagging vs.&nbsp;Random Forests</h2>
<ul>
<li>
<code>randomForest::randomForest(x, y, mtry, ntree, nodesize, sampsize)</code> <!-- + `mtry`: number of randomly sampled variable to consider at each internal node ($m$) --> <!-- + `ntree`: number of trees ($B$) --> <!-- + `nodesize`: stop splitting when the size of terminal nodes is no larger than nodesize --> <!-- + `sampsize`: how many samples to use when fitting each tree -->
</li>
</ul>
<div class="columns">
<div class="column" style="width:33%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-41-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:33%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-42-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:33%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-43-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div></div>
<aside class="notes"><p>nodesize: Minimum size of terminal nodes. Setting this number larger causes smaller trees to be grown sampsize: Size(s) of sample to draw. ntree: Number of trees to grow. mtry: Number of variables randomly sampled as candidates at each split.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside></section><section id="boosting" class="slide level2"><h2>Boosting</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><span class="green"><strong>Bagging</strong></span></p>
<ul>
<li><p>Trees are built on independent bootstrap data sets.</p></li>
<li><p>Trees are grown deep.</p></li>
<li><p>Large number of trees (<span class="math inline">\(B\)</span>) won’t overfit.</p></li>
</ul>
</div><div class="column" style="width:50%;">
<p><span class="green"><strong>Boosting</strong></span></p>
<ul>
<li><p>Trees are grown <em>sequentially</em>: each tree is grown using information from previously grown trees.</p></li>
<li><p>Each tree is fit on a <em>modified version of the original data set</em>, the residuals/false predictions!</p></li>
<li><p>Trees are rather small (weak learner).</p></li>
<li><p>Large <span class="math inline">\(B\)</span> can overfit.</p></li>
</ul>
</div></div>
<!-- - Bagging trees are built on bootstrap data sets, independent with each other. -->
<!-- - **Boosting** trees are grown *sequentially*: each tree is grown using information from previously grown trees. -->
<!-- - Boosting does not involve bootstrap sampling; instead each -->
<!-- tree is fit on a *modified version of the original data set*, the residuals/false predictions!  -->
</section><section id="boosting-1" class="slide level2"><h2>Boosting</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="xsmall">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="./images/14-tree/bagging_s24.png" style="width:100.0%"></p>
<figcaption>Source: https://pub.towardsai.net/bagging-vs-boosting-the-power-of-ensemble-methods-in-machine-learning-6404e33524e6</figcaption></figure>
</div>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="xsmall">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="./images/14-tree/boosting_s24.webp" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div>
</div></div>
<!-- - Consider producing a sequence of learners (trees), $\hat{f}^1, \dots, \hat{f}^B$. -->
<!-- - Boosting learns slowly  -->
<aside class="notes"><p>https://pub.towardsai.net/bagging-vs-boosting-the-power-of-ensemble-methods-in-machine-learning-6404e33524e6 - [Boosting can also be applied to many models, but most effectively applied to decision trees.] - Bagging trees are grown deep, but Boosting trees are rather small.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside></section><section id="boosting-for-classification" class="slide level2"><h2>Boosting for Classification</h2>
<ul>
<li>
<code>distribution = "bernoulli"</code>: <strong>LogitBoost</strong>
</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gbm.fit</span> <span class="op">=</span> <span class="fu">gbm</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gbm/man/gbm.html">gbm</a></span><span class="op">(</span><span class="va">y</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">x1</span>, <span class="va">x2</span>, <span class="va">y</span><span class="op">)</span>, </span>
<span>                   distribution <span class="op">=</span> <span class="st">"bernoulli"</span>, </span>
<span>                   n.trees <span class="op">=</span> <span class="fl">10000</span>, shrinkage <span class="op">=</span> <span class="fl">0.01</span>, bag.fraction <span class="op">=</span> <span class="fl">0.6</span>, </span>
<span>                   interaction.depth <span class="op">=</span> <span class="fl">2</span>, cv.folds <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="columns">
<div class="column" style="width:33%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-48-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:33%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-49-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:33%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img data-src="images/14-tree/unnamed-chunk-50-1.png" class="quarto-figure quarto-figure-center" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div></div>
<aside class="notes"><p>shrinkage: a shrinkage parameter applied to each tree in the expansion. bag.fraction: the fraction of the training set observations randomly selected to propose the next tree in the expansion. interaction.depth: Integer specifying the maximum depth of each tree n.trees: Integer specifying the total number of trees to fit.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside></section><section id="boosting-cross-validation" class="slide level2"><h2>Boosting Cross Validation</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">gbm.perf</span><span class="op">(</span><span class="va">gbm.fit</span>, method <span class="op">=</span> <span class="st">"cv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="cell-output cell-output-stdout">
<pre><code>[1] 1181</code></pre>
</div>
</div>
<img data-src="images/14-tree/unnamed-chunk-51-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:100.0%"></section><section id="boosting-for-regression" class="slide level2"><h2>Boosting for Regression</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gbm.fit</span> <span class="op">&lt;-</span> <span class="fu">gbm</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gbm/man/gbm.html">gbm</a></span><span class="op">(</span><span class="va">y</span> <span class="op">~</span> <span class="va">x</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span>, </span>
<span>                    distribution <span class="op">=</span> <span class="st">"gaussian"</span>, n.trees <span class="op">=</span> <span class="fl">300</span>,</span>
<span>                    shrinkage <span class="op">=</span> <span class="fl">0.5</span>, bag.fraction <span class="op">=</span> <span class="fl">0.8</span>, cv.folds <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<img data-src="images/14-tree/unnamed-chunk-53-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:100.0%"></section><section id="boosting-for-regression-1" class="slide level2"><h2>Boosting for Regression</h2>

<!-- ## Summary of Tree Ensemble Methods -->
<!-- - **Bagging** the trees are grown independently on random samples of -->
<!-- the observations. Consequently, the trees tend to be quite similar to -->
<!-- each other. Thus, bagging can get caught in local optima and can fail -->
<!-- to thoroughly explore the model space. -->
<!-- - **Random forests** the trees are once again grown independently on -->
<!-- random samples of the observations. However, each split on each tree -->
<!-- is performed using a random subset of the features, thereby decorrelating -->
<!-- the trees, and leading to a more thorough exploration of model -->
<!-- space relative to bagging. -->
<!-- - **Boosting**, we only use the original data, and do not draw any random -->
<!-- samples. The trees are grown successively, using a “slow” learning -->
<!-- approach: each new tree is fit to the signal that is left over from -->
<!-- the earlier trees, and shrunken down before it is used. -->
<img data-src="images/14-tree/unnamed-chunk-54-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:100.0%"></section><section id="other-topics" class="slide level2"><h2>Other Topics</h2>
<!-- - Boosting -->
<ul>
<li><p>AdaBoost (Adaptive Boosting) <code>gbm(y ~ ., distribution = "adaboost")</code></p></li>
<li>
<p>Gradient Boosting/Extreme Gradient Boosting (XGBoost) <code>xgboost</code></p>
<ul>
<li><a href="http://uc-r.github.io/gbm_regression" class="uri">http://uc-r.github.io/gbm_regression</a></li>
<li><a href="https://xgboost.readthedocs.io/en/stable/R-package/xgboostPresentation.html" class="uri">https://xgboost.readthedocs.io/en/stable/R-package/xgboostPresentation.html</a></li>
<li><a href="https://github.com/dmlc/xgboost" class="uri">https://github.com/dmlc/xgboost</a></li>
<li><a href="https://cran.r-project.org/web/packages/xgboost/xgboost.pdf" class="uri">https://cran.r-project.org/web/packages/xgboost/xgboost.pdf</a></li>
</ul>
</li>
<li><p>Bayesian Additive Regression Trees (BART) (ISL Sec. 8.2.4)</p></li>
</ul>
<aside class="notes"><p>https://jamleecute.web.app/gradient-boosting-machines-gbm/</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>

</section></section>
</div>
  <div class="quarto-auto-generated-content" style="display: none;">
<p><img src="https://upload.wikimedia.org/wikipedia/commons/e/e8/Marquette_Golden_Eagles_logo.svg" class="slide-logo"></p>
<div class="footer footer-default">
<p><a href="https://mssc6250-s25.github.io/website/">mssc6250-s25.github.io/website</a></p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="../site_libs/revealjs/dist/reveal.js"></script><!-- reveal.js plugins --><script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script><script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script><script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script><script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script><script src="../site_libs/revealjs/plugin/multiplex/socket.io.js"></script><script src="../site_libs/revealjs/plugin/multiplex/multiplex.js"></script><script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script><script src="../site_libs/revealjs/plugin/notes/notes.js"></script><script src="../site_libs/revealjs/plugin/search/search.js"></script><script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script><script src="../site_libs/revealjs/plugin/math/math.js"></script><script>window.define = window.backupDefine; window.backupDefine = undefined;</script><script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'multiplex': {"secret":"17441294780339158418","id":"3092197cab5a09b9","url":"https://reveal-multiplex.glitch.me/"},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1800,

        height: 1000,

        // Factor of the display size that should remain empty around the content
        margin: 5.0e-2,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script><script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script><script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp("https:\/\/mssc6250-s25\.github\.io\/website\/");
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>


</body></html>